packages/cip2/src/types.ts

```diff
@@ -0,0 +1,73 @@
+import { ProtocolParametersAlonzo } from '@cardano-ogmios/schema';
+import { TransactionUnspentOutput, Value, TransactionOutput } from '@emurgo/cardano-serialization-lib-nodejs';
+
+export interface SelectionResult {
+  selection: {
+    /**
+     * A set of inputs, equivalent to a subset of the initial UTxO set.
+     *
+     * From the point of view of a wallet, this represents the value
+     * that has been selected from the wallet in order to cover the total payment value.
+     */
+    inputs: TransactionUnspentOutput[];
+    /**
+     * Set of payments to be made to recipient addresses.
+     */
+    outputs: TransactionOutput[];
+    /**
+     * A set of change values. Does not account for fee.
+     *
+     * From the point of view of a wallet, this represents the change to be returned to the wallet.
+     */
+    change: Value[];
+    /**
+     * Estimated fee for the transaction
+     */
+    fee: bigint;
+  };
+  /**
+   * The remaining UTxO set is a subset of the initial UTxO set.
+   *
+   * It represents the set of values that remain after the coin selection algorithm
+   * has removed values to pay for entries in the requested output set.
+   */
+  remainingUTxO: TransactionUnspentOutput[];
+}
+
+export type EstimateTxFee = (
+  // Review: should we be using TransactionOutputs type instead?
+  // It's a collection object in CSL.
+  // Same question for InputSelection interface.
+  utxo: TransactionUnspentOutput[],
+  outputs: TransactionOutput[],
+  change: Value[]
+) => bigint;
```

Comment:
```suggestion
) => Promise<number>;
```

Reply:
We need to fully construct the transaction in order to estimate the fee, which requires a call to: https://github.com/input-output-hk/cardano-js-sdk/blob/e37fce63b00645897ae1ef6e0220c75ec133d163/packages/wallet/src/KeyManagement/KeyManager.ts#L12-L14

Reply:
That's unfortunate, it makes the entire `InputSelector.select` async

Reply:
We could also just change the interface, as the in-memory implementation isn't async:

https://github.com/input-output-hk/cardano-js-sdk/blob/master/packages/in-memory-key-manager/src/InMemoryKey.ts#L43-L50

I'll double-check to see what we need to support hardware wallets

Reply:
Unfortunately it's an async operation:

https://github.com/cardano-foundation/ledgerjs-hw-app-cardano/blob/47927d42e9b2b0a70463a719cbd745a84cbfce81/src/interactions/signTx.ts#L41-L56